### DISCUSSION TITLE
Cannot use Python Lib

### DISCUSSION THREAD

**Discussion by 01miaom** [beeware/beeware#117] [Q&A]
I want to be able to add a database in the Android APP, such as sqlite3.

I tried two methods. One is to add sqlite-android-3360000.aar to the libs folder, and then modify build.gradle. The APP can be successfully built, but the Android APP was crash.

Another way, I added sqlite directly to the "include" folder, and it doesn't seem to work.  Is there any good way to storing data in Android app?

**Comment by freakboy3742** [comment]
You should be able to use the built-in Python `sqlite` library with no more than `import sqlite` in your Python code - is that not working for you?

**Reply by 01miaom** [reply]
Thank for your reply! 
SQLite or other libraries can run well under the command of `briefcase dev`, but it fail to work after being packaged into Android. After preliminary testing, some API for file IO, such as `np. Load() np. Save(), open(), read(), write()` cannot run on Android, It may cause the app crash or fail to work. I guess it may be caused by the different file storage mechanism between the Linux and Android. So I try to move the database to the server and transfer data through FTP. I hope it can avoid strange and complex bugs like this.

**Reply by freakboy3742** [reply]
The Python test suite passes the sqlite tests, so it *can* work. If it's not working for you, then we need to debug *that*, rather than diving into how to replace the built-in SQLite.

Are you able to provide logs of the errors you're seeing?

**Comment by 01miaom** [comment]
I used the following code to test the sqlite function:
```
import toga
#import numpy as np
import os
import sqlite3 
from toga.style.pack import COLUMN, LEFT, RIGHT, ROW, Pack



def build(app):
    box = toga.Box()
    input0 = toga.TextInput()

    conn = sqlite3.connect('test.db') 
    cur = conn.cursor()
    
    def createTable(widget):
            sql_1 = "CREATE TABLE t(colume1 NUMBER)"
            cur.execute(sql_1)
            
    def insertData(widget):
            sql_2 = "INSERT INTO t VALUES(999)"
            cur.execute(sql_2)
                
    def getData(widget):
        try:  
            sql_3 = "SELECT * FROM t WHERE colume1==999"
            cur.execute(sql_3)
            #cur.fetchall()#search first line
            #print(cur.fetchone())
            a=cur.fetchone()#get search
            b=a.__str__()
            print(b[1:4])
            c = int(b[1:4])
            input0.value = c
            
        except ValueError:
            input0.value = '???'

    button0 = toga.Button('create table', on_press = createTable)
    button = toga.Button('get data', on_press = getData)
    button1 = toga.Button('insert data', on_press = insertData)
    box.add(input0)
    box.add(button0)
    box.add(button1)
    box.add(button)
    #input0.style.update(flex=1, padding_left=50)
    return box


def main():
    return toga.App('sqltest', 'test', startup=build)


if __name__ == '__main__':
    main().main_loop()
```
The program can run successfully under the `briefcase dev` command, but the program packaged with the `briefcase create android`  and 
 `briefcase build android` command cannot display the interface when tested on Android 10. The log of the build process is as follows:

**Reply by freakboy3742** [reply]
That's the build log, and it tells me that the app has successfully built. The *runtime* logs are what will tell me what is going wrong.

If I had to guess based on the code sample, the failure will be that you're loading "test.db", and it can't find that file because it's specified as a path relative to the current working directory. The current working directory on a dev app will be different to that on a packaged app. If you specify an absolute path (or, at least, a path relative to one of the source files), I'd bet you'll have more luck.
